In [1]:
import pandas as pd
import numpy as np

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
# from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report, confusion_matrix


from sklearn.metrics import RocCurveDisplay
from sklearn.metrics import roc_auc_score

In [2]:
x_train = pd.read_csv("../data/processed/x_train.csv")
x_test = pd.read_csv("../data/processed/x_test.csv")

y_train = pd.read_csv("../data/processed/y_train.csv")
y_test = pd.read_csv("../data/processed/y_test.csv")

## **Model Training**

### Logistic Regression with Grid Search 

In [3]:
log_reg = LogisticRegression(max_iter = 1000 , random_state=42)

param_grid = {
    'penalty': ['l1', 'l2'],
    'C': [0.01, 0.1, 1, 10, 50, 100],
    'class_weight': [None, 'balanced']
}


grid_search = GridSearchCV(
    estimator=log_reg,
    param_grid=param_grid,
    cv=5,
    scoring='accuracy',
    n_jobs=-1,
    verbose=1
)



In [6]:
models = {}
for trait in y_train.columns:
    grid_search.fit(x_train, y_train[trait])
    best_model = grid_search.best_estimator_
    models[trait] = best_model


for trait, model in models.items():

    y_pred_train = model.predict(x_train)
    y_pred_test = model.predict(x_test)
   
    y_pred = model.predict(x_test)
    print(f"Classification Report for {trait}:\n")
    print(classification_report(y_test[trait], y_pred))
    print(f"Confusion Matrix for {trait}:\n")
    print(confusion_matrix(y_test[trait], y_pred))
    print("Train Accuracy:", accuracy_score(y_train[trait], y_pred_train))
    print("Test Accuracy:", accuracy_score(y_test[trait], y_pred_test))
    print('\n')

Fitting 5 folds for each of 24 candidates, totalling 120 fits


c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\model_selection\_validation.py:516: FitFailedWarning: 
60 fits failed out of a total of 120.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
60 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\model_selection\_validation.py", line 859, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\base.py", line 1365, in wrapper
    return fit_method(estimator, *args, **kwargs)
  File "c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\linear_model\_logistic.py", line 1218

Fitting 5 folds for each of 24 candidates, totalling 120 fits


c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\model_selection\_validation.py:516: FitFailedWarning: 
60 fits failed out of a total of 120.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
60 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\model_selection\_validation.py", line 859, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\base.py", line 1365, in wrapper
    return fit_method(estimator, *args, **kwargs)
  File "c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\linear_model\_logistic.py", line 1218

Fitting 5 folds for each of 24 candidates, totalling 120 fits


c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\model_selection\_validation.py:516: FitFailedWarning: 
60 fits failed out of a total of 120.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
60 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\model_selection\_validation.py", line 859, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\base.py", line 1365, in wrapper
    return fit_method(estimator, *args, **kwargs)
  File "c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\linear_model\_logistic.py", line 1218

Fitting 5 folds for each of 24 candidates, totalling 120 fits


c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\model_selection\_validation.py:516: FitFailedWarning: 
60 fits failed out of a total of 120.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
60 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\model_selection\_validation.py", line 859, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\base.py", line 1365, in wrapper
    return fit_method(estimator, *args, **kwargs)
  File "c:\Users\rawan osama\anaconda3\envs\ml_env\lib\site-packages\sklearn\linear_model\_logistic.py", line 1218

Classification Report for Is_Extrovert:

              precision    recall  f1-score   support

           0       0.84      0.84      0.84      5930
           1       0.84      0.84      0.84      6070

    accuracy                           0.84     12000
   macro avg       0.84      0.84      0.84     12000
weighted avg       0.84      0.84      0.84     12000

Confusion Matrix for Is_Extrovert:

[[4965  965]
 [ 957 5113]]
Train Accuracy: 0.8397258276214088
Test Accuracy: 0.8398333333333333


Classification Report for Is_Sensor:

              precision    recall  f1-score   support

           0       0.76      0.78      0.77      5965
           1       0.78      0.76      0.77      6035

    accuracy                           0.77     12000
   macro avg       0.77      0.77      0.77     12000
weighted avg       0.77      0.77      0.77     12000

Confusion Matrix for Is_Sensor:

[[4634 1331]
 [1433 4602]]
Train Accuracy: 0.7624117169107689
Test Accuracy: 0.7696666666666667


Cl

### Logistic Regression Insights: 

**For Is_Sensor target:** 

Best Parameters: {'C': 0.1, 'class_weight': 'balanced', 'penalty': 'l2'}

Train Accuracy: 0.7624117169107689
Test Accuracy: 0.7696666666666667

---

**For Is_Extrovert target:**

Best Parameters: {'C': 0.01, 'class_weight': 'balanced', 'penalty': 'l2'}

Train Accuracy: 0.8397258276214088 / 
Test Accuracy: 0.8398333333333333

---

**For Is_Thinker target:**

Best Parameters: {'C': 0.1, 'class_weight': None, 'penalty': 'l2'}

Train Accuracy: 0.7930581887122649 / 
Test Accuracy: 0.794

---

**For Is_Judger target:**

Best Parameters: {'C': 10, 'class_weight': None, 'penalty': 'l2'}

Train Accuracy: 0.7798704139669577 / 
Test Accuracy: 0.77175


The results shows no overfittng nor underfitting , but it not the best accuracy 

---

--- 

### Decision Tree with Grid Search 

In [13]:
param_grid = {
    'max_depth': [3,5,7,10],
    'min_samples_split': [2,5,10],
    'min_samples_leaf': [1,2,5]
}

grid = GridSearchCV(DecisionTreeClassifier(random_state=42),
                    param_grid,
                    cv=5,
                    scoring='accuracy')



In [14]:
models = {}
for trait in y_train.columns:
    grid.fit(x_train, y_train[trait])
    best_dt = grid.best_estimator_
    models[trait] = best_dt


for trait, model in models.items():

    y_pred_train = model.predict(x_train)
    y_pred_test = model.predict(x_test)
   
    y_pred = model.predict(x_test)
    print(f"Classification Report for {trait}:\n")
    print(classification_report(y_test[trait], y_pred))
    print(f"Confusion Matrix for {trait}:\n")
    print(confusion_matrix(y_test[trait], y_pred))
    print("Train Accuracy:", accuracy_score(y_train[trait], y_pred_train))
    print("Test Accuracy:", accuracy_score(y_test[trait], y_pred_test))
    print('\n')

Classification Report for Is_Extrovert:

              precision    recall  f1-score   support

           0       0.82      0.85      0.84      5930
           1       0.85      0.82      0.84      6070

    accuracy                           0.84     12000
   macro avg       0.84      0.84      0.84     12000
weighted avg       0.84      0.84      0.84     12000

Confusion Matrix for Is_Extrovert:

[[5027  903]
 [1069 5001]]
Train Accuracy: 0.8838725806787642
Test Accuracy: 0.8356666666666667


Classification Report for Is_Sensor:

              precision    recall  f1-score   support

           0       0.82      0.83      0.82      5965
           1       0.83      0.82      0.82      6035

    accuracy                           0.82     12000
   macro avg       0.82      0.82      0.82     12000
weighted avg       0.82      0.82      0.82     12000

Confusion Matrix for Is_Sensor:

[[4931 1034]
 [1107 4928]]
Train Accuracy: 0.8721223358819975
Test Accuracy: 0.8215833333333333


Cl

### **Decision Tree Insights:**

**For Is_Sensor target:** 

Best Parameters: {'max_depth': 10, 'min_samples_leaf': 2, 'min_samples_split': 2}

Train Accuracy: 0.8721223358819975 / 
Test Accuracy: 0.8215833333333333

---

**For Is_Extrovert target:**

Best Parameters: {'C': 0.01, 'class_weight': 'balanced', 'penalty': 'l2'}

Train Accuracy: 0.8397258276214088 / 
Test Accuracy: 0.8398333333333333

---

**For Is_Thinker target:**

Best Parameters: {'max_depth': 10, 'min_samples_leaf': 2, 'min_samples_split': 2}

Train Accuracy: 0.8705806370966062 / 
Test Accuracy: 0.8255833333333333

---

**For Is_Judger target:**

Best Parameters: {'max_depth': 10, 'min_samples_leaf': 5, 'min_samples_split': 2}

Train Accuracy: 0.871497322860893 / 
Test Accuracy: 0.8269166666666666


---

**For Is_Judger target:**

Best Parameters: {'max_depth': 10, 'min_samples_leaf': 5, 'min_samples_split': 2}

Train Accuracy: 0.871497322860893 / 
Test Accuracy: 0.8269166666666666



The model shows a little overfitting , but it has better results than the logistic Regression 


---
---

#### **Random Forest**

In [16]:
models = {}
for trait in y_train.columns:
    rf_classifier = RandomForestClassifier(random_state=42)
    rf_classifier.fit(x_train, y_train[trait])
    models[trait] = rf_classifier


for trait, model in models.items():

    y_pred_train = model.predict(x_train)
    y_pred_test = model.predict(x_test)
   
    y_pred = model.predict(x_test)
    print(f"Classification Report for {trait}:\n")
    print(classification_report(y_test[trait], y_pred))
    print(f"Confusion Matrix for {trait}:\n")
    print(confusion_matrix(y_test[trait], y_pred))
    print("Train Accuracy:", accuracy_score(y_train[trait], y_pred_train))
    print("Test Accuracy:", accuracy_score(y_test[trait], y_pred_test))
    print("========================================================================\n")

Classification Report for Is_Extrovert:

              precision    recall  f1-score   support

           0       0.97      0.97      0.97      5930
           1       0.97      0.97      0.97      6070

    accuracy                           0.97     12000
   macro avg       0.97      0.97      0.97     12000
weighted avg       0.97      0.97      0.97     12000

Confusion Matrix for Is_Extrovert:

[[5760  170]
 [ 200 5870]]
Train Accuracy: 1.0
Test Accuracy: 0.9691666666666666

Classification Report for Is_Sensor:

              precision    recall  f1-score   support

           0       0.97      0.98      0.97      5965
           1       0.98      0.97      0.97      6035

    accuracy                           0.97     12000
   macro avg       0.97      0.97      0.97     12000
weighted avg       0.97      0.97      0.97     12000

Confusion Matrix for Is_Sensor:

[[5831  134]
 [ 199 5836]]
Train Accuracy: 1.0
Test Accuracy: 0.97225

Classification Report for Is_Thinker:

      

### **Random Forest Insights**: 


**For Is_Extrovert target:** 

Train Accuracy: 1.0 / 
Test Accuracy: 0.9691666666666666

---

**For Is_Sensor target:**

Train Accuracy: 1.0 / 
Test Accuracy: 0.97225

---

**For Is_Thinker target:**

Train Accuracy: 1.0 / 
Test Accuracy: 0.96358333

---

**For Is_Judger target:**

Train Accuracy: 1.0 / 
Test Accuracy: 0.9663

This model shows good results in both train and test accuracy , but there's a little bit overfitting 

---
---

#### **SVM**

In [15]:
models = {}
for trait in y_train.columns:
    svm_classifier = SVC(random_state=42)
    svm_classifier.fit(x_train, y_train[trait])
    models[trait] = svm_classifier
    
for trait, model in models.items():

    y_pred_train = model.predict(x_train)
    y_pred_test = model.predict(x_test)
   
    y_pred = model.predict(x_test)
    print(f"Classification Report for {trait}:\n")
    print(classification_report(y_test[trait], y_pred))
    print(f"Confusion Matrix for {trait}:\n")
    print(confusion_matrix(y_test[trait], y_pred))
    print("Train Accuracy:", accuracy_score(y_train[trait], y_pred_train))
    print("Test Accuracy:", accuracy_score(y_test[trait], y_pred_test))
    print("========================================================================\n")

Classification Report for Is_Extrovert:

              precision    recall  f1-score   support

           0       0.99      0.99      0.99      5930
           1       0.99      0.99      0.99      6070

    accuracy                           0.99     12000
   macro avg       0.99      0.99      0.99     12000
weighted avg       0.99      0.99      0.99     12000

Confusion Matrix for Is_Extrovert:

[[5884   46]
 [  54 6016]]
Train Accuracy: 0.9939373736952853
Test Accuracy: 0.9916666666666667

Classification Report for Is_Sensor:

              precision    recall  f1-score   support

           0       0.99      0.99      0.99      5965
           1       0.99      0.99      0.99      6035

    accuracy                           0.99     12000
   macro avg       0.99      0.99      0.99     12000
weighted avg       0.99      0.99      0.99     12000

Confusion Matrix for Is_Sensor:

[[5922   43]
 [  54 5981]]
Train Accuracy: 0.9936873684868435
Test Accuracy: 0.9919166666666667

Clas

### **SVM Insights:**

**For Is_Extrovert target:**

Train Accuracy: 0.9939 / 
Test Accuracy: 0.9916

---

**For Is_Sensor target:**

Train Accuracy: 0.9939 / 
Test Accuracy: 0.9916

---

**For Is_Thinker target:**

Train Accuracy: 0.9930 / 
Test Accuracy: 0.9904

---

**For Is_Judger target:**

Train Accuracy: 0.9931 / 
Test Accuracy: 0.99125

-------

**This model shows a perfect results in both training and test accuracy , showing no overfitting nor underfitting**


---
---

### **Model Insights Summary**

* **Logistic Regression:**
Showed stable performance with no overfitting, but accuracy was relatively lower than other models.

* **Decision Tree:**
Achieved better accuracy than Logistic Regression, with slight overfitting.

* **Random Forest:**
Gave very high accuracy, but perfect training score indicates some overfitting.

* **SVM:**
Achieved the best performance with very high and consistent train/test accuracy, showing excellent generalization.


#### **Final Choice**

The **SVM model** was selected as the final model because it achieved the highest accuracy with no overfitting, making it the most reliable for this task.